iothub/device/src/ModuleClient.cs

```diff
@@ -359,6 +373,88 @@ public void SetRetryPolicy(IRetryPolicy retryPolicy)
+        }
+
+        /// <summary>
+        /// Interactively invokes a method on module
+        /// </summary>
+        /// <param name="deviceId">Device Id</param>
+        /// <param name="moduleId">Module Id</param>
+        /// <param name="directMethodRequest">Device method parameters (passthrough to device)</param>
+        /// <returns>Method result</returns>
+        public Task<DirectMethodResult> InvokeMethodAsync(string deviceId, string moduleId, DirectMethodRequest directMethodRequest)
+        {
+            return this.InvokeMethodAsync(deviceId, moduleId, directMethodRequest, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Interactively invokes a method on module
+        /// </summary>
+        /// <param name="deviceId">Device Id</param>
+        /// <param name="moduleId">Module Id</param>
+        /// <param name="directMethodRequest">Device method parameters (passthrough to device)</param>
+        /// <param name="cancellationToken">Cancellation Token</param>
+        /// <returns>Method result</returns>
+        public Task<DirectMethodResult> InvokeMethodAsync(string deviceId, string moduleId, DirectMethodRequest directMethodRequest, CancellationToken cancellationToken)
+        {
+            return InvokeMethodAsync(GetModuleMethodUri(deviceId, moduleId), directMethodRequest, cancellationToken);
+        }
+
+        private Task<DirectMethodResult> InvokeMethodAsync(Uri uri, DirectMethodRequest directMethodRequest, CancellationToken cancellationToken)
+        {
+            var httpClientHandler = new HttpClientHandler();
+
+#if NETSTANDARD1_3 || NETSTANDARD2_0
+            httpClientHandler.ServerCertificateCustomValidationCallback = this.certValidator.GetCustomCertificateValidation();
+#endif
+
+            var context = new PipelineContext();
+            context.Set(new ProductInfo() { Extra = this.InternalClient.ProductInfo });
+
+            Http1TransportSettings transportSettings = new Http1TransportSettings();
+            //We need to add the certificate to the httpTransport if DeviceAuthenticationWithX509Certificate
+            if (this.internalClient.Certificate != null)
+            {
+                transportSettings.ClientCertificate = this.internalClient.Certificate;
+            }
+
+            HttpTransportHandler httpTransport = new HttpTransportHandler(context, this.internalClient.IotHubConnectionString, transportSettings, httpClientHandler);
+            return httpTransport.InvokeMethodAsync(directMethodRequest, uri, cancellationToken);
+        }
+
+        static Uri GetDeviceMethodUri(string deviceId)
+        {
+            deviceId = WebUtility.UrlEncode(deviceId);
+            return new Uri(string.Format(CultureInfo.InvariantCulture, DeviceMethodUriFormat, deviceId), UriKind.Relative);
+        }
+
+        static Uri GetModuleMethodUri(string deviceId, string moduleId)
+        {
+            deviceId = WebUtility.UrlEncode(deviceId);
```

Comment:
url encode moduleId

Reply:
done, also added in ServiceClient

---
In reply to: [193912221](https://github.com/Azure/azure-iot-sdk-csharp/pull/501#discussion_r193912221) [](ancestors = 193912221)